brainiak/searchlight/searchlight.py

```diff
@@ -0,0 +1,378 @@
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import numpy as np
+from mpi4py import MPI
+import sys
+
+"""Distributed Searchlight
+"""
+
+__all__ = [
+        "Searchlight",
+]
+
+
+class Searchlight:
+    """Distributed Searchlight
+
+    Run a user-defined function over each voxel in a multi-subject
+    dataset.
+
+    Optionally, users can define a block function which runs over
+    larger portions of the volume called blocks.
+    """
+    def __init__(self, sl_rad=1, max_blk_edge=10):
+        """Constructor
+
+        Parameters
+        ----------
+
+        sl_rad: radius, in voxels, of the sphere inscribed in the
+                   searchlight cube, not counting the center voxel
+
+        max_blk_edge: max edge length, in voxels, of the 3D block
+
+        """
+        self.sl_rad = sl_rad
+        self.max_blk_edge = max_blk_edge
+        self.comm = MPI.COMM_WORLD
+
+    def _get_ownership(self, data):
+        """Determine on which rank each subject currently resides
+
+        Parameters
+        ----------
+
+        data: list of 4D arrays with subject data
+
+        Returns
+        -------
+
+        list of ranks indicating the owner of each subject
+        """
+        rank = self.comm.rank
+
+        B = [(rank, idx) for (idx, c) in enumerate(data) if c is not None]
+        C = self.comm.allreduce(B)
+        ownership = [None] * len(data)
+        for c in C:
+            ownership[c[1]] = c[0]
+        return ownership
+
+    def _get_blocks(self, mask):
+        """Divide the volume into a set of blocks
+
+        Ignore blocks that have no active voxels in the mask
+
+        Parameters
+        ----------
+
+        mask: a boolean 3D array which is true at every active voxel
+
+        Returns
+        -------
+
+        list of tuples containing block information:
+           - a triple containing top left point of the block and
+           - a triple containing the size in voxels of the block
+
+        """
+        outerblk = self.max_blk_edge + 2*self.sl_rad
+        return [((i, j, k),
+                mask[i:i+outerblk,
+                j:j+outerblk,
+                k:k+outerblk].shape)
+                for i in range(0, mask.shape[0], self.max_blk_edge)
+                for j in range(0, mask.shape[1], self.max_blk_edge)
+                for k in range(0, mask.shape[2], self.max_blk_edge)
+                if np.any(
+                mask[i+self.sl_rad:i+self.sl_rad+self.max_blk_edge,
+                     j+self.sl_rad:j+self.sl_rad+self.max_blk_edge,
+                     k+self.sl_rad:k+self.sl_rad+self.max_blk_edge])]
+
+    def _get_block_data(self, mat, block):
+        """Retrieve a block from a 3D or 4D volume
+
+        Parameters
+        ----------
+
+        mat: a 3D or 4D volume
+
+        block: a tuple containing block information:
+          - a triple containing the top left point of the block and
```

Comment:
Top left front or top left back? Also, wouldn't it be better to remove direction altogether and say "the voxel with the lowest coordinates"?